src/utils/cuda/cuda_utils.cpp

```diff
@@ -0,0 +1,369 @@
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "cuda_utils.h"
+#include <nixl_log.h>
+
+#ifdef HAVE_CUDA
+
+#include <cuda_runtime.h>
+#include <cuda.h>
+
+#endif
+
+
+// TODO: remove:
+// #define HAVE_CUDA 1
+// #define  HAVE_CUMEMRETAINALLOCATIONHANDLE 1
+// #define HAVE_DECL_CU_MEM_LOCATION_TYPE_HOST 1
+
+
+/****************************************
+ * CUDA nixlCudaPtr class
+*****************************************/
+
+#ifdef HAVE_CUDA
+
+class nixlCudaMemCtxImpl : public nixlCudaMemCtx {
+private:
+    CUcontext ctx;
+
+    nixl_status_t queryVmm(const void *address, memory_t &type, uint64_t &id);
+    nixl_status_t retainVmmCudaCtx(uint64_t id, CUcontext &newCtx) const;
+    nixl_status_t releaseVmmCudaCtx(uint64_t id) const;
+    nixl_status_t queryCuda(const void *address, memory_t &type, uint64_t &id,
+                            CUcontext &newCtx);
+
+public:
+
+    nixlCudaMemCtxImpl() : nixlCudaMemCtx()
+    {    }
+
+    ~nixlCudaMemCtxImpl() override {
+        if (MEM_VMM_DEV == memType) {
+            nixl_status_t status = releaseVmmCudaCtx(_devId);
+            if (NIXL_SUCCESS != status) {
+                NIXL_ERROR << "Failed to release CUDA context";
+            }
+        }
+    }
+
+    nixl_status_t enableAddr(const void *address, uint64_t chkDevId) override;
+
+    nixl_status_t set() override;
+};
+
+#endif
+
+/****************************************
+ * Static nixlCudaPtr functions
+*****************************************/
+
+#ifdef HAVE_CUDA
+
+#define NIXL_CUDA_PTR_CTX_CLASS nixlCudaMemCtxImpl
```

Comment:
can we move HAVE_CUDA usage inside the factory function, and avoid defining NIXL_CUDA_PTR_CTX_CLASS macro for simplicity?